tests/suite/math/mat.typ

```diff
@@ -54,13 +54,64 @@ $ a + mat(delim: #none, 1, 2; 3, 4) + b $
   $ mat(1, 2; 3, 4; delim: "[") $,
 )
 
+--- math-mat-spread ---
+// Test argument spreading in matrix.
+$ mat(..#range(1, 5).chunks(2))
+  mat(#(..range(2).map(_ => range(2)))) $
+
+#let nums = ((1,) * 5).intersperse(0).chunks(3)
+$ mat(..nums, delim: "[") $
+
+--- math-mat-spread-complex ---
+// Test argument spreading in matrix in complex situations.
+$ mat(delim: "|",
+  1, 2, ..., 10;
+  ..#range(0, 1).map(i => range(i, i + 4)),
+  ..#range(0, 1).map(i => range(i, i + 4)),
+  1, 2, ..., 10,
+) $
+
+--- math-mat-spread-expected-array-error-1 ---
+// Error: 13-14 expected array, found content
+$ mat(1, 2; 3, ..#range(0, 2)) $
+
+--- math-mat-spread-expected-array-error-2 ---
+#let nums = range(0, 2).map(i => (i, i+1))
+// Error: 20 expected array, found content
+$ mat(1, 2; ..nums;, delim: #none) $
+
```

Comment:
I'm not sure these tests match desired behavior. I feel like if we have spreads, they should always merge with surrounding arguments (and two adjacent spreads merge with each-other), requiring explicit semicolons to create separate rows in the matrix.

Playing around with 2d-arguments, I have a difficult time building a mental model of how normal arguments interact with spreads under the current system.

With normal args, changing a middle comma to a semicolon (i.e. `$mat(1,2,3,4)$` -> `$mat(1,2;3,4)$`) works fine, but spread arguments only join into one row when there are no semicolons. Adding a semicolon requires completely reshaping the data to avoid producing an error. In: `$mat(..#($1$,$2$), ..#($3$,$4$))$` -> `$mat(..#($1$,$2$) ; ..#($3$,$4$))$` -- the second equation gives: "Error: expected array, found content" on the _first_ argument.

This behavior feels wrong to me and feels like a reason to re-evaluate how spread arguments and semicolons should operate together.

This does also impact spreading arguments that create multiple arrays like below. I'm not sure what would be preferable here.
```typ
$ mat(..#range(1, 5).chunks(2)) $
```

I'll note that my current personal draft of math runtime parsing merges spreads with normal arguments. It's not worth waiting for that to land just to get this feature, but it will be a problem if the behavior changes between the two. I'd rather that we make this behavior as correct as possible here without runtime parsing.

I don't have a simple way to resolve this just within the parser. It could be that we need to add a new `MathArgs` Kind and AST node to get the correct behavior from the AST?

Opinions? @PgBiel @Enivex

Reply:
I think we can get this more expected behaviour quite simply by treating spread arguments as positional. The behaviour is a lot clearer then I think:

Spreading 1d arrays merges with other positional arguments as expected, since the result of spreading this should be no different than just writing the elements of the array directly. For example,
```typ
mat(..#range(1, 5) ; 1, ..#range(2, 5)) = mat(1, 2, 3, 4 ; 1, 2, 3, 4)
mat(..#range(1, 3), ..#range(3, 5) ; ..#range(1, 4), 4) = mat(1, 2, 3, 4 ; 1, 2, 3, 4)
```

Spreading a 2d array as the only argument still works:
```typ
mat(..#((1,) * 5).intersperse(0).chunks(3), delim: "[") = mat(1, 0, 1; 0, 1, 0; 1, 0, 1; delim: "[")
```

Spreading a 2d array with other positional arguments and semicolons are a little trickier at first, but still make sense with the behaviour a user will know from what was presented before. For example:
```typ
#let nums = range(0, 2).map(i => (i, i+1))

// Will display (0,1) and (1,2) verbatim as
// using the semicolon re-wraps it into an array
$ mat(..nums;) $

// Will throw an "expected array, found content"
// error as the end empty argument isn't in an array
// (same as what happens currently on main)
$ mat(1, 2; ..nums;, delim: #none) $

// Will throw an "expected array, found content"
// error as the last two arguments aren't in an
// array
$ mat(..nums, 0, 1) $

// Will throw an "expected array, found integer"
// error as it is equivalent to `(..nums, 0, 1)`
$ mat(..nums, ..#range(0, 2))
```

More clearly, here is the repr of some arguments from the tests:
```typ
#let nums = range(0, 4).chunks(2)
#check($args(..nums)$, "((0, 1), (2, 3))")
#check($args(..nums;)$, "(((0, 1), (2, 3)))")
#check($args(..nums, ..nums)$, "((0, 1), (2, 3), (0, 1), (2, 3))")
#check($args(..nums, 4, 5)$, "((0, 1), (2, 3), [4], [5])")
#check($args(..nums, ..#range(4, 6))$, "((0, 1), (2, 3), 4, 5)")
#check($args(..nums, #range(4, 6))$, "((0, 1), (2, 3), (4, 5))")
#check($args(..nums, 1, 2; 3, 4)$, "(((0, 1), (2, 3), [1], [2]), ([3], [4]))")
#check($args(1, 2; ..nums)$, "(([1], [2]), ((0, 1), (2, 3)))")
#check($args(1, 2; 3, 4)$, "(([1], [2]), ([3], [4]))")
#check($args(1, 2; 3, 4; ..#range(5, 7))$, "(([1], [2]), ([3], [4]), (5, 6))")
#check($args(1, 2; 3, 4, ..#range(5, 7))$, "(([1], [2]), ([3], [4], 5, 6))")
#check($args(1, 2; 3, 4, ..#range(5, 7);)$, "(([1], [2]), ([3], [4], 5, 6))")
#check($args(1, 2; 3, 4, ..#range(5, 7),)$, "(([1], [2]), ([3], [4], 5, 6))")
```